File: src/main/webapp/src/app/github/service/GithubPrCriteria.ts
import {GitHubPrCriteriaDTO, GitHubPullRequestDTO} from '../../rest';
import {Crit} from '../../utils/Crit';

export class GithubPrCriteria {

  static anyCriteriaSet(c: GitHubPrCriteriaDTO): boolean {
    return c.fromYear != null
      || c.toYear != null
      || (c.usernamePattern ?? '').trim().length > 0
      || c.fromPullRequestNumber != null
      || c.toPullRequestNumber != null
      || (c.pullRequestNumberPattern ?? '').trim().length > 0
      || Crit.parseCsvList(c.titleContains).length > 0
      || Crit.parseCsvList(c.bodyContains).length > 0
      || Crit.parseCsvList(c.authorContains).length > 0
      || Crit.parseCsvList(c.labelContains).length > 0
      || Crit.parseCsvList(c.baseRefContains).length > 0
      || Crit.parseCsvList(c.excludedStates).length > 0
      || (c.draftAvailability ?? 'any') !== 'any'
      || (c.mergedAvailability ?? 'any') !== 'any'
      || (c.mergeableAvailability ?? 'any') !== 'any'
      || (c.mergeableStatePattern ?? '').trim().length > 0
      || Crit.parseCsvList(c.analysisSummaryContains).length > 0
      || (c.analysisSummaryUpdatedFrom ?? '').length > 0
      || (c.analysisSummaryUpdatedTo ?? '').length > 0
      || c.analysisSummaryMinTokensK != null
      || c.analysisSummaryMaxTokensK != null
      || Crit.parseCsvList(c.analysisUserExtraPromptsContains).length > 0
      || (c.analysisAvailability ?? 'any') !== 'any'
      || Crit.parseCsvList(c.developmentWorkDescribedContains).length > 0
      || (c.developmentWorkUpdatedFrom ?? '').length > 0
      || (c.developmentWorkUpdatedTo ?? '').length > 0
      || c.developmentWorkMinTokensK != null
      || c.developmentWorkMaxTokensK != null
      || Crit.parseCsvList(c.developmentWorkUserExtraPromptsContains).length > 0
      || (c.developmentWorkAvailability ?? 'any') !== 'any'
      || Crit.parseCsvList(c.personalInterrestCommentContains).length > 0
      || c.personalInterrestMinPriority != null
      || c.personalInterrestMaxPriority != null
      || (c.personalInterrestAvailability ?? 'any') !== 'any';
  }

  static match(c: GitHubPrCriteriaDTO, pr: GitHubPullRequestDTO): boolean {
    if (!Crit.matchesDateTextInYearRange(c.fromYear, c.toYear, pr.createdAt)) {
      return false;
    }
    if (c.usernamePattern) {
      const name = (pr.authorLogin && pr.authorLogin.trim()) ? pr.authorLogin : 'unknown';
      if (!Crit.matchesRegex(c.usernamePattern, name)) {
        return false;
      }
    }
    if (!Crit.matchesRegex(c.pullRequestNumberPattern, pr.number != null ? String(pr.number) : undefined)) {
      return false;
    }
    if (!Crit.matchesNumberRange(c.fromPullRequestNumber, c.toPullRequestNumber, pr.number)) {
      return false;
    }
    if (!Crit.matchesAny(c.titleContains, pr.title)) {
      return false;
    }
    if (!Crit.matchesAny(c.bodyContains, pr.body)) {
      return false;
    }
    if (!Crit.matchesAny(c.authorContains, pr.authorLogin, ...(pr.assigneeLogins ?? []), ...(pr.requestedReviewerLogins ?? []))) {
      return false;
    }
    if (!Crit.matchesAny(c.labelContains, ...(pr.labelNames ?? []))) {
      return false;
    }
    if (!Crit.matchesAny(c.baseRefContains, pr.baseRef)) {
      return false;
    }
    if (pr.state != null && Crit.isExcluded(c.excludedStates, pr.state)) {
      return false;
    }
    if (!Crit.matchesYesNoAny(c.draftAvailability, !!pr.draft)) {
      return false;
    }
    if (!Crit.matchesYesNoAny(c.mergedAvailability, !!pr.merged)) {
      return false;
    }
    if (!Crit.matchesYesNoAny(c.mergeableAvailability, !!pr.mergeable)) {
      return false;
    }
    if (!Crit.matchesRegex(c.mergeableStatePattern, pr.mergeableState)) {
      return false;
    }
    const annotated = pr.annotated;
    if (!Crit.matchesYesNoAny(c.analysisAvailability, !!annotated?.analysisSummary)) {
      return false;
    }
    if (!Crit.matchesAny(c.analysisSummaryContains, annotated?.analysisSummary)) {
      return false;
    }
    if (!Crit.matchesDateRange(c.analysisSummaryUpdatedFrom, c.analysisSummaryUpdatedTo, annotated?.analysisSummaryLastUpdateTime)) {
      return false;
    }
    if (!Crit.matchesTokensRangeK(c.analysisSummaryMinTokensK, c.analysisSummaryMaxTokensK, annotated?.analysisSummaryTokensConsumed)) {
      return false;
    }
    if (!Crit.matchesAny(c.analysisUserExtraPromptsContains, ...(annotated?.analysisUserExtraPrompts ?? []))) {
      return false;
    }
    if (!Crit.matchesYesNoAny(c.developmentWorkAvailability, !!annotated?.developmentWorkDescribed)) {
      return false;
    }
    if (!Crit.matchesAny(c.developmentWorkDescribedContains, annotated?.developmentWorkDescribed)) {
      return false;
    }
    if (!Crit.matchesDateRange(c.developmentWorkUpdatedFrom, c.developmentWorkUpdatedTo, annotated?.developmentWorkLastUpdateTime)) {
      return false;
    }
    if (!Crit.matchesTokensRangeK(c.developmentWorkMinTokensK, c.developmentWorkMaxTokensK, annotated?.developmentWorkTokensConsumed)) {
      return false;
    }
    if (!Crit.matchesAny(c.developmentWorkUserExtraPromptsContains, ...(annotated?.developmentWorkUserExtraPrompts ?? []))) {
      return false;
    }
    if (!Crit.matchesYesNoAny(c.personalInterrestAvailability, !!annotated?.personalInterrestComment)) {
      return false;
    }
    if (!Crit.matchesAny(c.personalInterrestCommentContains, annotated?.personalInterrestComment)) {
      return false;
    }
    if (!Crit.matchesNumberRange(c.personalInterrestMinPriority, c.personalInterrestMaxPriority, annotated?.personalInterrestPriority10)) {
      return false;
    }
    return true;
  }

}

File: src/main/java/fr/an/projectanalysis/github/service/GitHubPrCriteria.java
package fr.an.projectanalysis.github.service;

import fr.an.projectanalysis.github.rest.dtos.GitHubPrCriteriaDTO;
import fr.an.projectanalysis.github.rest.dtos.GitHubPullRequestDTO;
import fr.an.projectanalysis.github.rest.dtos.GitHubPullRequestExtraFieldsDTO;
import fr.an.projectanalysis.util.CritUtils;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Predicate;

/**
 * Whether a {@link GitHubPullRequestDTO} matches the Main/Analysis/Development Work/Personal
 * Interest filter criteria of the github-pull-requests page (a null criteria matches everything).
 */
public class GitHubPrCriteria implements Predicate<GitHubPullRequestDTO> {

    private final GitHubPrCriteriaDTO c;

    public GitHubPrCriteria(GitHubPrCriteriaDTO c) {
        this.c = c;
    }

    @Override
    public boolean test(GitHubPullRequestDTO pr) {
        if (c == null) {
            return true;
        }
        if (!CritUtils.matchesAny(c.titleContains, pr.title)) {
            return false;
        }
        if (!CritUtils.matchesAny(c.bodyContains, pr.body)) {
            return false;
        }
        List<String> authorValues = new ArrayList<>();
        authorValues.add(pr.authorLogin);
        if (pr.assigneeLogins != null) {
            authorValues.addAll(pr.assigneeLogins);
        }
        if (pr.requestedReviewerLogins != null) {
            authorValues.addAll(pr.requestedReviewerLogins);
        }
        if (!CritUtils.matchesAny(c.authorContains, authorValues.toArray(String[]::new))) {
            return false;
        }
        List<String> labels = pr.labelNames != null ? pr.labelNames : List.of();
        if (!CritUtils.matchesAny(c.labelContains, labels.toArray(String[]::new))) {
            return false;
        }
        if (!CritUtils.matchesAny(c.baseRefContains, pr.baseRef)) {
            return false;
        }
        if (CritUtils.isExcluded(c.excludedStates, pr.state)) {
            return false;
        }
        if (!CritUtils.matchesAvailability(c.draftAvailability, pr.draft)) {
            return false;
        }
        if (!CritUtils.matchesAvailability(c.mergedAvailability, pr.merged)) {
            return false;
        }
        if (!CritUtils.matchesAvailability(c.mergeableAvailability, Boolean.TRUE.equals(pr.mergeable))) {
            return false;
        }
        if (!CritUtils.matchesRegex(c.mergeableStatePattern, pr.mergeableState)) {
            return false;
        }

        GitHubPullRequestExtraFieldsDTO annotated = pr.annotated;
        boolean hasAnalysis = annotated != null && annotated.analysisSummary != null && !annotated.analysisSummary.isBlank();
        if (!CritUtils.matchesAvailability(c.analysisAvailability, hasAnalysis)) {
            return false;
        }
        if (!CritUtils.matchesAny(c.analysisSummaryContains, annotated != null ? annotated.analysisSummary : null)) {
            return false;
        }
        if (!CritUtils.matchesDateRange(c.analysisSummaryUpdatedFrom, c.analysisSummaryUpdatedTo,
                annotated != null ? annotated.analysisSummaryLastUpdateTime : null)) {
            return false;
        }
        if (!CritUtils.matchesTokensRangeK(c.analysisSummaryMinTokensK, c.analysisSummaryMaxTokensK,
                annotated != null ? annotated.analysisSummaryTokensConsumed : 0)) {
            return false;
        }
        List<String> analysisExtraPrompts = annotated != null && annotated.analysisUserExtraPrompts != null
                ? annotated.analysisUserExtraPrompts : List.of();
        if (!CritUtils.matchesAny(c.analysisUserExtraPromptsContains, analysisExtraPrompts.toArray(String[]::new))) {
            return false;
        }

        boolean hasDevWork = annotated != null && annotated.developmentWorkDescribed != null && !annotated.developmentWorkDescribed.isBlank();
        if (!CritUtils.matchesAvailability(c.developmentWorkAvailability, hasDevWork)) {
            return false;
        }
        if (!CritUtils.matchesAny(c.developmentWorkDescribedContains, annotated != null ? annotated.developmentWorkDescribed : null)) {
            return false;
        }
        if (!CritUtils.matchesDateRange(c.developmentWorkUpdatedFrom, c.developmentWorkUpdatedTo,
                annotated != null ? annotated.developmentWorkLastUpdateTime : null)) {
            return false;
        }
        if (!CritUtils.matchesTokensRangeK(c.developmentWorkMinTokensK, c.developmentWorkMaxTokensK,
                annotated != null ? annotated.developmentWorkTokensConsumed : 0)) {
            return false;
        }
        List<String> devWorkExtraPrompts = annotated != null && annotated.developmentWorkUserExtraPrompts != null
                ? annotated.developmentWorkUserExtraPrompts : List.of();
        if (!CritUtils.matchesAny(c.developmentWorkUserExtraPromptsContains, devWorkExtraPrompts.toArray(String[]::new))) {
            return false;
        }

        boolean hasPersonalInterrest = annotated != null && annotated.personalInterrestComment != null && !annotated.personalInterrestComment.isBlank();
        if (!CritUtils.matchesAvailability(c.personalInterrestAvailability, hasPersonalInterrest)) {
            return false;
        }
        if (!CritUtils.matchesAny(c.personalInterrestCommentContains, annotated != null ? annotated.personalInterrestComment : null)) {
            return false;
        }
        if (!CritUtils.matchesNumberRange(c.personalInterrestMinPriority, c.personalInterrestMaxPriority,
                annotated != null ? annotated.personalInterrestPriority10 : null)) {
            return false;
        }
        return true;
    }

}
